indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedResourceManagementStrategy.java

```diff
@@ -0,0 +1,360 @@
+    }
+    want = Math.min(want, maxWorkerCount - currValidWorkers);
+    return want;
+  }
+
+  int getWorkersNeededToAssignTasks(
+      final RemoteTaskRunnerConfig remoteTaskRunnerConfig,
+      final WorkerBehaviorConfig workerConfig,
+      final Collection<Task> pendingTasks,
+      final Collection<ZkWorker> zkWorkers
+  )
+  {
+    final Collection<ZkWorker> validWorkers = Collections2.filter(
+        zkWorkers,
+        ResourceManagementUtil.createValidWorkerPredicate(config)
+    );
+
+    Map<String, ImmutableZkWorker> workersMap = Maps.newHashMap();
+    for (ZkWorker worker : validWorkers) {
+      workersMap.put(worker.getWorker().getHost(), worker.toImmutable());
+    }
+    WorkerSelectStrategy workerSelectStrategy = workerConfig.getSelectStrategy();
+    int need = 0;
+    int capacity = getAverageWorkerCapacity(zkWorkers);
+    for (Task task : pendingTasks) {
+      Optional<ImmutableZkWorker> selectedWorker = workerSelectStrategy.findWorkerForTask(
+          remoteTaskRunnerConfig,
+          ImmutableMap.copyOf(workersMap),
+          task
+      );
+      final ImmutableZkWorker workerRunningTask;
+      if (selectedWorker.isPresent()) {
+        workerRunningTask = selectedWorker.get();
+      } else {
+        // None of the existing worker can run this task, we need to provision one worker for it.
+        need++;
+        workerRunningTask = createDummyWorker("dummy" + need, capacity, remoteTaskRunnerConfig.getMinWorkerVersion());
+      }
+      // Update map with worker running task
+      workersMap.put(workerRunningTask.getWorker().getHost(), workerWithTask(workerRunningTask, task));
+    }
+    return need;
+  }
+
+  @Override
+  public boolean doTerminate(RemoteTaskRunner runner)
+  {
+    Collection<ZkWorker> zkWorkers = runner.getWorkers();
+    synchronized (lock) {
+      final WorkerBehaviorConfig workerConfig = workerConfigRef.get();
+      if (workerConfig == null) {
+        log.warn("No workerConfig available, cannot terminate workers.");
+        return false;
+      }
+
+      boolean didTerminate = false;
+      final Set<String> workerNodeIds = Sets.newHashSet(
+          workerConfig.getAutoScaler().ipToIdLookup(
+              Lists.newArrayList(
+                  Iterables.transform(
+                      runner.getLazyWorkers(),
+                      new Function<ZkWorker, String>()
+                      {
+                        @Override
+                        public String apply(ZkWorker input)
+                        {
+                          return input.getWorker().getIp();
+                        }
+                      }
+                  )
+              )
+          )
+      );
+
+      final Set<String> stillExisting = Sets.newHashSet();
+      for (String s : currentlyTerminating) {
+        if (workerNodeIds.contains(s)) {
+          stillExisting.add(s);
+        }
+      }
+      currentlyTerminating.clear();
+      currentlyTerminating.addAll(stillExisting);
```

Comment:
wouldn't it be simpler to just do the following inside the loop rather than creating a separate collection?
```
if (!workerNodeIds.contain(s)) { currentlyTerminating.remove(s) }
```